### ISSUE TITLE
Item details redesign

### ISSUE THREAD

**Issue by nielsvanvelzen** [jellyfin/jellyfin-androidtv#269]
In an effort to fix legacy code and make Jellyfin looks nice we are going to revamp the detail screen. This screen currently contains code to show details for tv shows, episodes, movies, actors, people, artists and some more and the code is a huge mess.

A discussion emerged and our current goal is to use the androidx.leanback library as a base and add custom rows for Jellyfin related data. Everything will be written in Kotlin.

We decided to make the changes in small steps in a newly created branch. The initial PR for this is #270. When everything is ready it will be merged to master. Progress will be posted in the comments.

## Team
Current members:
- @thornbill - reviewing
- @nielsvanvelzen - programming, reviewing
Previous members:
- @AndreasGB - programming, reviewing

## Issues
_These will be updated from time to time to track all issues related to the redesign_
- #270 
- #275 
- #276 
- #278 
- #285 
- #295 
- #307
- #321

## References
- https://designguidelines.withgoogle.com/android-tv/tv-apps/detail-view.html
- https://developer.android.com/jetpack/androidx/releases/leanback
- https://developer.android.com/reference/androidx/leanback/app/package-summary

**Comment by AndreasGB** [comment]
As discussed on matrix, we should probably next tackle getting working buttons and a background.

Probably it would make sense to create a `BaseAction` subclass from Action and have all actions be further children of it, and then add a custom `OnActionClickedListener`, which on calls some handler on the actions. This would allow for re-use of actions in different DetailFragments, e.g. `Resume` would be needed for movies as well as TV episodes.

**Comment by nielsvanvelzen** [comment]
Sounds like a good plan. I'd like to point out that #251 also contains actions so we could probably take an example of @Florianisme their implementation for a bit.

**Comment by AndreasGB** [comment]
To improve playback buttons and other dealing with different item types, we will need an object oriented representation of BaseItemDto, and cannot continue with data objects. We should probably build those and have a conversion method that converts `BaseItemDto` to those new objects.

**Comment by AndreasGB** [comment]
Now that #276 has merged and we have the basics for a new item model nailed down, we can now start on implementing more item types and fragments.

**Comment by nielsvanvelzen** [comment]
I think we should focus on adding one type with as much content as possible. So all images, ratings, descriptions, related items and such. We can then use that as a reference.

I'd suggest using the Movie type for it because it doesn't have relations to other types (Like TVSeries<>Seasons<>Episodes or MusicAlbum<>MusicArtist)

**Comment by AndreasGB** [comment]
We also have a sneak peek at the current progress which is of course very much not finished (captured from the demo server).

![Progress](https://user-images.githubusercontent.com/8351447/73126118-30df7900-3faf-11ea-93f8-3523ef394094.png)

**Comment by AndreasGB** [comment]
Re @nielsvanvelzen: Sounds like a better idea to avoid duplication of work.

**Comment by AndreasGB** [comment]
As some of the interactions with this UI lead to changes in the items, a mechanism is needed to observe changes on the items for synchronization between UI and data. For example, setting items watched should make the resume button disappear. @nielsvanvelzen suggested using [LiveData](https://developer.android.com/topic/libraries/architecture/livedata)

**Comment by nielsvanvelzen** [comment]
After #285 is merged the following "big" things need to be done:
- Create custom presenter for DetailsOverviewRow with our own xml layout
- Implement custom rows
- Add missing item types (currently only movie/episode)

After that's done we can start cleaning the code and test everything thoroughly.

**Comment by nextlooper42** [comment]
Is part of redesign also options to set audio and subtitle track **before** clicking on play? It would save a lot of time on server side and also some flustration on user time (less waiting). System now is kinda annoying, you need to play movie/show and wait if to load then hit pause, click on audio what i want and click on subs what i want and wait them to "render", which takes a lot of time

**Comment by nielsvanvelzen** [comment]
I want to add that and already started on it in #307 (look at the screenshot in my latest comment)

**Comment by nielsvanvelzen** [comment]
I did some searching to find out which item types use the FullDetailsActivity but it's not that easy to find out. I got multiple lists to get an idea which items _could_ be displayed:

## BaseItemTypes references in FullDetailsActivity (excluding buttonTypes):
 - BaseItemType.Episode
 - BaseItemType.Series
 - BaseItemType.Person
 - BaseItemType.MusicArtist
 - BaseItemType.Program
 - BaseItemType.Recording
 - BaseItemType.Movie

## FullDetailsActivity.buttonTypes property:
 - BaseItemType.Episode
 - BaseItemType.Movie
 - BaseItemType.Series
 - BaseItemType.Season
 - BaseItemType.Folder
 - BaseItemType.Video
 - BaseItemType.Recording
 - BaseItemType.Program
 - BaseItemType.ChannelVideoItem
 - BaseItemType.Trailer
 - BaseItemType.MusicArtist
 - BaseItemType.Person
 - BaseItemType.MusicVideo
 - BaseItemType.SeriesTimer

## FullDetailsActivity.addAdditionalRows switch case:
 - BaseItemType.Movie
 - BaseItemType.Trailer
 - BaseItemType.Person
 - BaseItemType.MusicArtist
 - BaseItemType.Series
 - BaseItemType.Episode
 - BaseItemType.SeriesTimer

## ItemLauncher references to FullDetailsActivity for:
 - BaseItemType.Series
 - BaseItemType.MusicArtist
 - BaseItemType.Person
 - BaseItemType.SearchHint - with a condition
 - BaseItemType.LiveTvProgram
 - BaseItemType.LiveTvRecording
 - BaseItemType.SeriesTimer

**Comment by nielsvanvelzen** [comment]
Because I wasn't satisfied with the architecture I changed a lot of it in #368. Unfortunately I came to the conclusion that Leanback (the "row" system in it to be more specific) doesn't really like data changing after the views have been created. I still need to think of a proper way to fix this. In that branch the actions don't work properly (they do not update visually).

Another issue I found was the fact that we used the constructor of the fragments to pass data. This is not allowed in Android as the fragment needs to have an empty constructor that is used when resuming an activity. To fix this we need to send the data in the bundle in the `onCreate` function. This causes the next issue where our data classes need to implement the `Parcelable` interface, which can't be done nicely because we don't use data classes, which we do not use because we wanted to inherit properties. One way to solve this would be to only pass the item-id to the fragment (in the bundle) and have it load from a repository. This repository would use the API and cache to retrieve the information. This does take quite some time to implement though.

**Comment by nielsvanvelzen** [comment]
Pinning for awareness

**Comment by AndreasGB** [comment]
I'm going to implement `EpisodeDetailsFragment` next

**Comment by AndreasGB** [comment]
As this is now merged, I will do `SeasonDetailsFragment` next, starting in the evening.

**Comment by nielsvanvelzen** [comment]
I'm going to work on the music related items (currently MusicArtist)

**Comment by nextlooper42** [comment]
First of all, thank you for all your work on this. I really appreciate it.

I've been testing item branch (18.6.2020) to test out my translation, among the other thigs on Sony 4k tv (ATV3). Apart from Android Studio emulated view is real life hw slightly different, mainly in size of things. Here are my findings about redesign so far (i know that it is unfinished)... i mean problems to solve:

![now](https://user-images.githubusercontent.com/29764734/85081034-1e1e4880-b1cb-11ea-845f-e3dbfa3aed3b.png)

- Logo in right upper corner is way too big and too close to edges on screen, it does not look good. It need to be little smaller and more centered somehow.
- There is way too much stuff on screen
- Whole details on right side should not be there when you open item, it doesn't fit to screen when you have more then one thing... and even that one thing looks messy (eg. audio text stretches to 3 lines...). Imagine how subs looks when i have minumum 10 different languages as audios and 25 different subtitles in mkv.
- Title panel is way too short, it doesnt fit most of longer names
- Tags should be somewhere lower on detail page
- Description panel should be bigger
- Year and Genres should be in one line
- Chapter selection line is the biggest thing on page, and it makes no sense. It needs to be way smaller.
- Move duration and end time to second page (last screen of this post) to tag bar
- Whole item detail is kinda slow and lagging (especially when clicked down to chapter/details list)
- Do we need icons in Play/Watched/etc row? I dont recall any other service to have them. And it makes translated string to break.
- Switch actor photos to circle maybe?

![page12](https://user-images.githubusercontent.com/29764734/85081857-7f471b80-b1cd-11ea-848b-9c9a9c850ea0.jpg)

I havent tested tv show item redesign, since it always crash the app.

Design should be much more simpler than what is there now. Follow more https://designguidelines.withgoogle.com/android-tv/tv-apps/detail-view.html or just take a look at Disney+, Netflix or Hulu and take a page from their book. Less is more in the end and it will look cleaner & be faster (srsly).

**Comment by AndreasGB** [comment]
Thanks for the detailed feedback. Circle-cropping the photos has been tried by us and it does not work as there are many photos where the actor is off-center. Of course, one could use face detection and try to crop by that, but that's way too much of a hassle for little visual benefit.
